R/bitbucket.R

```diff
@@ -0,0 +1,138 @@
+# Copy of code from install_github.R.
+# release tag is not supported by Bitbucket
+parse_bitbucket_repo <- function(path) {
+  username_rx <- "(?:([^/]+)/)?"
+  repo_rx <- "([^/@#]+)"
+  subdir_rx <- "(?:/([^@#]*[^@#/]))?"
+  ref_rx <- "(?:@([^*].*))"
+  pull_rx <- "(?:#([0-9]+))"
+  ref_or_pull_rx <- sprintf("(?:%s|%s)?", ref_rx, pull_rx)
+  bitbucket_rx <- sprintf("^(?:%s%s%s%s|(.*))$",
+    username_rx, repo_rx, subdir_rx, ref_or_pull_rx)
+
+  param_names <- c("username", "repo", "subdir", "ref", "pull", "invalid")
+  replace <- stats::setNames(sprintf("\\%d", seq_along(param_names)), param_names)
```

Comment:
I think rather than duplicating this logic it would be better to have a function which takes a path, regex and names and returns the result that can be used both here and in `parse_git_repo()`.

```r

parse_repo <- function(path, rx, nms) {
   replace <- stats::setNames(sprintf("\\%d", seq_along(nms)), nms)
  params <- lapply(replace, function(r) gsub(rx, r, path, perl = TRUE))
  if (params$invalid != "")
    stop(sprintf("Invalid repo: %s", path))
  params <- params[sapply(params, nchar) > 0]

  params
}